packages/flutter/lib/src/widgets/shared_state.dart

```diff
@@ -0,0 +1,181 @@
+///
+/// This widget doesn't keep the state itself, it uses the state passed in. It
+/// calls the [valueChanged] callback when the state changes. You should then
+/// pass the updated state to this widget on the next build as the [value].
+///
+/// When the shared state's [value] is updated, all of the widgets that depend
+/// upon it will be rebuilt.
+///
+/// ## Sample code
+///
+/// ```dart
+/// enum Thing {
+///   raindrops,
+///   whiskers,
+///   kettles,
+///   mittens,
+///   kittens,
+///   packages,
+///   string,
+/// }
+/// 
+/// class FavoritePage extends StatefulWidget {
+///   const FavoritePage({Key key}) : super(key: key);
+/// 
+///   @override
+///   _FavoritePageState createState() => new _FavoritePageState();
+/// }
+/// 
+/// class _FavoritePageState extends State<FavoritePage> {
+///   Thing _favoriteThing;
+/// 
+///   @override
+///   Widget build(BuildContext context) {
+///     return new SharedState<Thing>(
+///       value: _favoriteThing,
+///       valueChanged: (Thing thing) {
+///         if (_favoriteThing != thing) {
+///           setState(() {
+///             _favoriteThing = thing;
+///           });
+///         }
+///       },
+///       // For this example, assume we don't wish to, or can't, pass
+///       // _favoriteThing to FavoriteView.
+///       child: const FavoriteView(),
+///     );
+///   }
+/// }
+/// 
+/// class FavoriteView extends StatelessWidget {
+///   const FavoriteView();
+/// 
+///   @override
+///   Widget build(BuildContext context) {
+///    return new ListView(
+///       children: Thing.values.map<Widget>((Thing thing) {
+///         return FavoriteTile(thing);
+///       }).toList(),
+///     );
+///   }
+/// }
+/// 
+/// class FavoriteTile extends StatelessWidget {
+///   const FavoriteTile(this.thing);
+/// 
+///   final Thing thing;
+/// 
+///   @override
+///   Widget build(BuildContext context) {
+///     final Thing favoriteThing = SharedState.get<Thing>(context);
+///     final bool isSelected = favoriteThing == thing;
+///     return ListTile(
+///       leading: new IconButton(
+///         icon: new Icon(isSelected ? Icons.favorite : Icons.favorite_border),
+///         onPressed: () {
+///           SharedState.set<Thing>(context, isSelected ? null : thing);
+///         },
+///       ),
+///       title: new Text(describeEnum(thing)),
+///     );
+///   }
+/// }
+/// ```
+class SharedState<T> extends InheritedWidget {
+  /// Creates a [SharedState].
+  ///
+  /// The `child` and `valueChanged` arguments are required and must not be null.
+  const SharedState({
+    Key key,
+    this.value,
+    @required this.valueChanged,
+    @required Widget child,
+  })  : assert(valueChanged != null),
+        assert(child != null),
+        super(key: key, child: child);
+
+  /// The value of the shared state.
+  final T value;
+
+  /// The [ValueChanged] callback that is called when the shared state is
+  /// changed.
+  final ValueChanged<T> valueChanged;
```

Comment:
How does this get called?  who calls it?

I think it makes more sense for SharedState to "listen" to the <T> and call valuechanged itself when the T says it has changed.